### PR TITLE
[Fixes #6533] Log when XML formatters fail to create a serializer

### LINKED ISSUE #6533: Log when XML formatters fail to create a serializer
See: https://github.com/aspnet/Mvc/issues/6527

We've also had similar issues reported in the past with DCS formatter. 

What's discussed in #6527 is by design. The XML formatters swallow any and all exceptions that come from creating a serializer as part of the content negotiation process. When a user doesn't know this, we have to ask the user to write 4-5 lines of code just to figure out what's going wrong. 

We should at a minimum add logging to the XML formatters to report this kind of usage as a warning. Now that the XML formatters are optional (off by default) it's much more likely that if you are using XML, then you intend all of your DTO's to be serializable by your XML formatter.

The current behavior was designed for a world where the XML formatters were on by default, but most users didn't really care about XML support. We wanted them to stay out of the way, but they are currently too far away. We could consider changing this in a 3.0.0 release and surfacing the exception.

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/MvcXmlDataContractSerializerMvcOptionsSetup.cs [hunk 1]
```diff
@@ -21,7 +39,9 @@ public void Configure(MvcOptions options)
         {
             options.ModelMetadataDetailsProviders.Add(new DataMemberRequiredBindingMetadataProvider());
 
-            options.OutputFormatters.Add(new XmlDataContractSerializerOutputFormatter());
+            var outputFormatterLogger = _loggerFactory.CreateLogger<XmlDataContractSerializerOutputFormatter>();
+
+            options.OutputFormatters.Add(new XmlDataContractSerializerOutputFormatter(outputFormatterLogger));
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/MvcXmlDataContractSerializerMvcOptionsSetup.cs]: This is strange. Where else do we create a type-specific logger for another type? I'd expect the `XmlDataContractSerializerOutputFormatter` constructor to get the `ILoggerFactory`.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerOutputFormatter.cs [hunk 2]
```diff
@@ -21,14 +22,25 @@ namespace Microsoft.AspNetCore.Mvc.Formatters
...
-        public XmlDataContractSerializerOutputFormatter() :
-            this(FormattingUtilities.GetDefaultXmlWriterSettings())
+        public XmlDataContractSerializerOutputFormatter()
+            : this(FormattingUtilities.GetDefaultXmlWriterSettings())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="XmlDataContractSerializerOutputFormatter"/>
+        /// with default XmlWriterSettings
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerOutputFormatter.cs]: End all comments with periods.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerOutputFormatter.cs [hunk 3]
```diff
@@ -37,6 +49,16 @@ public class XmlDataContractSerializerOutputFormatter : TextOutputFormatter
...
+            : this(writerSettings, logger: null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="XmlDataContractSerializerOutputFormatter"/>
+        /// </summary>
+        /// <param name="writerSettings">The settings to be used by the <see cref="DataContractSerializer"/>.</param>
+        /// <param name="logger">The <see cref="ILogger"/>.</param>
+        public XmlDataContractSerializerOutputFormatter(XmlWriterSettings writerSettings, ILogger logger)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerOutputFormatter.cs]: Other than tests, where are the constructor overloads accepting a `XmlWriterSettings` expected to be used? I'm wondering if we can limit the additions to one overload per formatter type.

File: test/Microsoft.AspNetCore.Mvc.Formatters.Xml.Test/XmlDataContractSerializerOutputFormatterTest.cs [hunk 4]
```diff
@@ -626,6 +628,58 @@ public void GetSupportedContentTypes_ReturnsSupportedTypes(Type type, object exp
             XmlAssert.Equal(expectedOutput, content);
         }
 
+        public static IEnumerable<object[]> LogsWhenUnableToCreateSerializerForTypeData
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Formatters.Xml.Test/XmlDataContractSerializerOutputFormatterTest.cs]: Would be slightly easier to understand using `TheoryData<XmlDataContractSerializerOutputFormatter, TestSink>`.

**Comment by kichalla** [comment]
🆙 📅

**Comment by kichalla** [comment]
05d02e7cab3744e9225959c6340d0a62e230f7a2

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerOutputFormatter.cs [hunk 5]
```diff
@@ -21,22 +22,43 @@ namespace Microsoft.AspNetCore.Mvc.Formatters
...
 
         /// <summary>
         /// Initializes a new instance of <see cref="XmlDataContractSerializerOutputFormatter"/>
-        /// with default XmlWriterSettings
+        /// with default <see cref="XmlWriterSettings"/>.
         /// </summary>
-        public XmlDataContractSerializerOutputFormatter() :
-            this(FormattingUtilities.GetDefaultXmlWriterSettings())
+        public XmlDataContractSerializerOutputFormatter()
+            : this(FormattingUtilities.GetDefaultXmlWriterSettings())
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerOutputFormatter.cs]: <3

**Pull request closed without merging**